homeassistant/components/zha/sensor.py

```diff
@@ -1367,3 +1368,164 @@ class AqaraCurtainHookStateSensor(EnumSensor):
+class BitMapSensor(Sensor):
+    """A sensor with only state attributes.
+
+    The sensor value will be a sensor of the state attributes.
+    """
+
+    _default_value: str
+    _bitmap: dict[str, int]
+
+    def formatter(self, _value: int) -> str:
+        """Summary of all attributes."""
+        binary_state_attributes = [
+            key for (key, elem) in self.extra_state_attributes.items() if elem
+        ]
+
+        return (
+            ", ".join(binary_state_attributes)
+            if binary_state_attributes
+            else self._default_value
+        )
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Bitmap."""
+        value = self._cluster_handler.cluster.get(self._attribute_name)
+
+        state_attr = {}
+
+        for text, bit in self._bitmap.items():
+            state_attr[text] = bool(value & bit)
+
+        return state_attr
+
+
+class DanfossOpenWindowDetectionEnum(types.enum8):
+    """Danfoss open window detection judgments."""
+
+    Quarantine = 0x00
+    Closed = 0x01
+    Maybe = 0x02
+    Open = 0x03
+    External = 0x04
+
+
+@MULTI_MATCH(
+    cluster_handler_names=CLUSTER_HANDLER_THERMOSTAT,
+    quirk_ids={DANFOSS_ALLY_THERMOSTAT},
+)
+# pylint: disable-next=hass-invalid-inheritance # needs fixing
+class DanfossOpenWindowDetection(EnumSensor):
+    """Danfoss proprietary attribute.
+
+    Sensor that displays whether the TRV detects an open window using the temperature sensor.
+    """
+
+    _unique_id_suffix = "open_window_detection"
+    _attribute_name = "open_window_detection"
+    _attr_translation_key: str = "open_window_detected"
+    _attr_icon: str = "mdi:window-open"
+    _enum = DanfossOpenWindowDetectionEnum
+
+
+@CONFIG_DIAGNOSTIC_MATCH(
+    cluster_handler_names=CLUSTER_HANDLER_THERMOSTAT,
+    quirk_ids={DANFOSS_ALLY_THERMOSTAT},
+)
+# pylint: disable-next=hass-invalid-inheritance # needs fixing
+class DanfossLoadEstimate(Sensor):
+    """Danfoss proprietary attribute for communicating its estimate of the radiator load."""
+
+    _unique_id_suffix = "load_estimate"
+    _attribute_name = "load_estimate"
+    _attr_translation_key: str = "load_estimate"
+    _attr_icon: str = "mdi:scale-balance"
+    _attr_entity_category = EntityCategory.DIAGNOSTIC
+
+
+@CONFIG_DIAGNOSTIC_MATCH(
+    cluster_handler_names=CLUSTER_HANDLER_THERMOSTAT,
+    quirk_ids={DANFOSS_ALLY_THERMOSTAT},
+)
+# pylint: disable-next=hass-invalid-inheritance # needs fixing
+class DanfossAdaptationRunStatus(BitMapSensor):
+    """Danfoss proprietary attribute for showing the status of the adaptation run."""
+
+    _unique_id_suffix = "adaptation_run_status"
+    _attribute_name = "adaptation_run_status"
+    _attr_translation_key: str = "adaptation_run_status"
+    _attr_entity_category = EntityCategory.DIAGNOSTIC
+    _default_value = "Nothing"
+    _bitmap = {
+        "In Progress": 0x0001,
+        "Run Successful": 0x0002,
+        "Valve characteristic lost": 0x0004,
+    }
```

Comment:
Hmm, it's not ideal that these aren't translated.
Also, the capitalization needs to be changed to match the HA standard (for all BitMapSensors): e.g. "In progress", ..

Reply:
Thanks, missed those capitals. Wasn't there a PR that mentioned the translation of enums (https://github.com/home-assistant/core/pull/96553 was closed)? I take it that that hasn't been done yet. Maybe I could take a look at that. The translation of enums could be done in a separate PR and the translation of bitmaps here as it is (for now) only needed for Danfoss.

Reply:
The extra_state_attributes can be easily translated for the bitmap, but the state of the entity doesn't seem to work for some reason. But if I can get that to work, translation of enums is also going to be easy.

Reply:
Okay, looked at the code for a couple of hours to realize the translations didn't get regenerated. It worked after all. I added every enum in strings.json. However for enums to work, new merge requests are required. I could also manually change the camelcase to snake case in my own enums, but that would make them inconsistent with zigpy.